### PR TITLE
Use median timestamp if current time renders a block invalid

### PR DESCRIPTION
**Pull request by thaerkh**
This addresses issue https://github.com/monero-project/monero/issues/3717

### CODE REVIEW

**Comment by stoffu** [comment]
How about this alternative?
```diff
diff --git a/src/cryptonote_core/blockchain.cpp b/src/cryptonote_core/blockchain.cpp
index 25e97f71f..9e51287f0 100644
--- a/src/cryptonote_core/blockchain.cpp
+++ b/src/cryptonote_core/blockchain.cpp
@@ -1142,6 +1142,12 @@ bool Blockchain::create_block_template(block& b, const account_public_address& m
   b.prev_id = get_tail_id();
   b.timestamp = time(NULL);
 
+  uint64_t median_ts;
+  if (!check_block_timestamp(b, median_ts))
+  {
+    b.timestamp = median_ts;
+  }
+
   diffic = get_difficulty_for_next_block();
   CHECK_AND_ASSERT_MES(diffic, false, "difficulty overhead.");
 
@@ -3127,10 +3133,10 @@ uint64_t Blockchain::get_adjusted_time() const
 }
 //------------------------------------------------------------------
 //TODO: revisit, has changed a bit on upstream
-bool Blockchain::check_block_timestamp(std::vector<uint64_t>& timestamps, const block& b) const
+bool Blockchain::check_block_timestamp(std::vector<uint64_t>& timestamps, const block& b, uint64_t& median_ts) const
 {
   LOG_PRINT_L3("Blockchain::" << __func__);
-  uint64_t median_ts = epee::misc_utils::median(timestamps);
+  median_ts = epee::misc_utils::median(timestamps);
 
   if(b.timestamp < median_ts)
   {
@@ -3148,7 +3154,7 @@ bool Blockchain::check_block_timestamp(std::vector<uint64_t>& timestamps, const
 //   true if the block's timestamp is not less than the timestamp of the
 //       median of the selected blocks
 //   false otherwise
-bool Blockchain::check_block_timestamp(const block& b) const
+bool Blockchain::check_block_timestamp(const block& b, uint64_t& median_ts) const
 {
   LOG_PRINT_L3("Blockchain::" << __func__);
   if(b.timestamp > get_adjusted_time() + CRYPTONOTE_BLOCK_FUTURE_TIME_LIMIT)
@@ -3173,7 +3179,7 @@ bool Blockchain::check_block_timestamp(const block& b) const
     timestamps.push_back(m_db->get_block_timestamp(offset));
   }
 
-  return check_block_timestamp(timestamps, b);
+  return check_block_timestamp(timestamps, b, median_ts);
 }
 //------------------------------------------------------------------
 void Blockchain::return_tx_to_pool(std::vector<transaction> &txs)
diff --git a/src/cryptonote_core/blockchain.h b/src/cryptonote_core/blockchain.h
index 61096fb26..66d25cc7c 100644
--- a/src/cryptonote_core/blockchain.h
+++ b/src/cryptonote_core/blockchain.h
@@ -1292,10 +1292,12 @@ namespace cryptonote
      *   false otherwise
      *
      * @param b the block to be checked
+     * @param median_ts return-by-value the median of timestamps
      *
      * @return true if the block's timestamp is valid, otherwise false
      */
-    bool check_block_timestamp(const block& b) const;
+    bool check_block_timestamp(const block& b, uint64_t& median_ts) const;
+    bool check_block_timestamp(const block& b) const { uint64_t median_ts; return check_block_timestamp(b, median_ts); }
 
     /**
      * @brief checks a block's timestamp
@@ -1305,10 +1307,12 @@ namespace cryptonote
      *
      * @param timestamps a list of the most recent timestamps to check against
      * @param b the block to be checked
+     * @param median_ts return-by-value the median of timestamps
      *
      * @return true if the block's timestamp is valid, otherwise false
      */
-    bool check_block_timestamp(std::vector<uint64_t>& timestamps, const block& b) const;
+    bool check_block_timestamp(std::vector<uint64_t>& timestamps, const block& b, uint64_t& median_ts) const;
+    bool check_block_timestamp(std::vector<uint64_t>& timestamps, const block& b) const { uint64_t median_ts; return check_block_timestamp(timestamps, b, median_ts); }
 
     /**
      * @brief get the "adjusted time"
```
I prefer this because the relationship among different versions of functions is well expressed in the header.

**Comment by iamsmooth** [comment]
> return-by-value

Isn't this return by reference?

**Comment by stoffu** [comment]
@iamsmooth 
You're right, thanks for correcting.

**Comment by moneromooo-monero** [comment]
Why the do not merge tag ? This seems OK. Did you intend to make more changes ?
If not, could you please also PR it to the release-0.12 branch, I think it should go in there too.

**Comment by stoffu** [comment]
@moneromooo-monero 

What's your opinion about my proposed alternative, which hasn't been reflected?

**Comment by stoffu** [comment]
@thaerkh 
> I'll make the style changes when I include some extra additions I'd like to have, and might have to merge it into #2887 (includes a protocol change)

I'm not sure what you mean by "some extra additions" and how this is related to #2887.

**Comment by moneromooo-monero** [comment]
The two patches seem pretty much identical to me. Am I missing a clever hack ?

**Comment by thaerkh** [comment]
@moneromooo-monero they're logically equivalent for intent, except that @stoffu  's alternative proposal includes the FTL check and my implementation segregated it to just the median's.

I've included the alternative proposal to accommodate coding convention. Since it includes the FTL check, I've also changed the block template to get the max of the median or top block's timestamp, to better handle bad clocks.

the "extra additions" from the other PR include protocol changes and would change this slightly, but this PR is isolated to not have any of that so this is probably sufficient.

Cheers,
Thaer

File: src/cryptonote_core/blockchain.cpp [hunk 1]
```diff
@@ -1142,6 +1142,12 @@ bool Blockchain::create_block_template(block& b, const account_public_address& m
   b.prev_id = get_tail_id();
   b.timestamp = time(NULL);
 
+  uint64_t median_ts;
+  if (!check_block_timestamp(b, median_ts))
+  {
+    b.timestamp = std::max(median_ts, m_db->get_top_block_timestamp());
```
Comment (moneromooo-monero) [inline src/cryptonote_core/blockchain.cpp]: That seems dangerous, no ? It means the miner of that block gets an inordinate say over what subsequent miners will timestamp with.

Comment (thaerkh) [reply src/cryptonote_core/blockchain.cpp]: Would you prefer the following? `b.timestamp = std::max(median_ts, m_db->get_top_block_timestamp() - CRYPTONOTE_BLOCK_FUTURE_TIME_LIMIT);`

Comment (moneromooo-monero) [reply src/cryptonote_core/blockchain.cpp]: I'd rather use b.timestamp = median_ts; as this is the smallest "sanity" change which makes the block pass consensus rules.

Comment (thaerkh) [reply src/cryptonote_core/blockchain.cpp]: That's reasonable - I've changed it back to `b.timestamp = median_ts;`

**Pull request merged**